### ISSUE TITLE
[Bug]: gemma3 shows degraded accuracy in vLLM v0.8.4

### ISSUE THREAD

**Issue by llsj14** [vllm-project/vllm#17689]
### Your current environment

<details>
<summary>The output of <code>python collect_env.py</code></summary>

```text
Your output of `python collect_env.py` here
```

</details>

### 🐛 Describe the bug

a

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by DarkLight1337** [comment]
Are you using V1 Engine? The attention mask of this model is simplified when using V1 Engine which may lead to some inaccuracies.

**Comment by llsj14** [comment]
@DarkLight1337 
Thank you for your quick advice. I actually tested with the v0 option as well, but it showed the same degraded results.
Below is the script I used to run lm_eval.

```
export HF_MODEL_PATH="/hugginface/gemma-3-27b-it"
export TP_SIZE=4
export TASKS=lambada_openai,hellaswag
export DTYPE=auto
export BATCH_SIZE=4

# VLLM
VLLM_USE_V1=0 VLLM_WORKER_MULTIPROC_METHOD=spawn CUDA_VISIBLE_DEVICES=0,1,2,3 python -m lm_eval --model=vllm \
     --model_args=pretrained=$HF_MODEL_PATH,tensor_parallel_size=$TP_SIZE,gpu_memory_utilization=0.75,enforce_eager=True,trust_remote_code=True,max_gen_toks=256 \
     --tasks=$TASKS \
     --batch_size=$BATCH_SIZE \
     --output_path=./outputs
```

**Comment by DarkLight1337** [comment]
@WoosukKwon any idea?

**Comment by airlsyn** [comment]
we met the same degraded result on gemma-3-27b-it with vllm==0.8.5.

**Comment by ashgold** [comment]
Is this okay for versions of VLLM below V0.8.4?

**Comment by airlsyn** [comment]
> Is this okay for versions of VLLM below V0.8.4?

v0.8.3 is okay.

**Comment by DarkLight1337** [comment]
Can you try whether setting `dtype="bfloat16"` explicitly fixes the issue?

**Comment by llsj14** [comment]
I ran the test using the bfloat16 option, but the results appear to be the same.
```
vllm (pretrained=/hugginface/gemma-3-27b-it,tensor_parallel_size=4,gpu_memory_utilization=0.75,enforce_eager=True,trust_remote_code=True,max_gen_toks=256,dtype=bfloat16), gen_kwargs: (None), limit: None, num_fewshot: None, batch_size: 4
```

|    Tasks     |Version|Filter|n-shot|  Metric  |   | Value  |   |Stderr |
|--------------|------:|------|-----:|----------|---|-------:|---|------:|
|hellaswag     |      1|none  |     0|acc       |↑  |  0.3859|±  | 0.0049|
|              |       |none  |     0|acc_norm  |↑  |  0.4859|±  | 0.0050|
|lambada_openai|      1|none  |     0|acc       |↑  |  0.2267|±  | 0.0058|
|              |       |none  |     0|perplexity|↓  |947.2092|±  |77.1147|

**Comment by DarkLight1337** [comment]
Hmm, can you test the text-only (1B) version of the model? It would greatly narrow down the scope of investigation.

**Comment by llsj14** [comment]
yeah, I tested `gemma-3-1b-it` model and the results were as follows.

### vLLM result
#### dtype auto
|    Tasks     |Version|Filter|n-shot|  Metric  |   | Value  |   |Stderr |
|--------------|------:|------|-----:|----------|---|-------:|---|------:|
|hellaswag     |      1|none  |     0|acc       |↑  |  0.3512|±  | 0.0048|
|              |       |none  |     0|acc_norm  |↑  |  0.4345|±  | 0.0049|
|lambada_openai|      1|none  |     0|acc       |↑  |  0.2307|±  | 0.0059|
|              |       |none  |     0|perplexity|↓  |226.0222|±  |14.3887|

#### dtype bfloat16
|    Tasks     |Version|Filter|n-shot|  Metric  |   | Value  |   |Stderr |
|--------------|------:|------|-----:|----------|---|-------:|---|------:|
|hellaswag     |      1|none  |     0|acc       |↑  |  0.3512|±  | 0.0048|
|              |       |none  |     0|acc_norm  |↑  |  0.4345|±  | 0.0049|
|lambada_openai|      1|none  |     0|acc       |↑  |  0.2307|±  | 0.0059|
|              |       |none  |     0|perplexity|↓  |226.0222|±  |14.3887|

### HF result
#### dtype auto
|    Tasks     |Version|Filter|n-shot|  Metric  |   | Value |   |Stderr|
|--------------|------:|------|-----:|----------|---|------:|---|-----:|
|hellaswag     |      1|none  |     0|acc       |↑  | 0.4348|±  |0.0049|
|              |       |none  |     0|acc_norm  |↑  | 0.5782|±  |0.0049|
|lambada_openai|      1|none  |     0|acc       |↑  | 0.4343|±  |0.0069|
|              |       |none  |     0|perplexity|↓  |28.4342|±  |1.4753|

**Comment by DarkLight1337** [comment]
We didn't change Gemma3 code much recently. apart from #17180.. @LucasWilkinson any idea what might cause this?

**Comment by llsj14** [comment]
>We encountered the same degraded results (almost zero) on gemma-3-27b-it using vLLM version 0.8.5.

hello, @airlsyn 
I'm considering the possibility that my test script might be incorrect, so I wanted to check if you also encountered this issue while using lm-evaluation-harness.

**Comment by LucasWilkinson** [comment]
this appears to be ok on main, @llsj14 do you mine testing to see if the issue still exists on the nightly?
```
|    Tasks     |Version|Filter|n-shot|  Metric  |   |Value |   |Stderr|
|--------------|------:|------|-----:|----------|---|-----:|---|-----:|
|hellaswag     |      1|none  |     0|acc       |↑  |0.6511|±  |0.0048|
|              |       |none  |     0|acc_norm  |↑  |0.8401|±  |0.0037|
|lambada_openai|      1|none  |     0|acc       |↑  |0.7046|±  |0.0064|
|              |       |none  |     0|perplexity|↓  |3.7754|±  |0.1154|
```

**Comment by llsj14** [comment]
@LucasWilkinson Thank you for sharing.
I tested again with the most recent vLLM main (last commit ID: 0c0fdae84f1da5e45518aafc7b32e8139055adae) and the latest lm-evaluation-harness, but the results were still the same, as follows:

vllm (pretrained=/checkpoints/hugginface/gemma-3-27b-it,tensor_parallel_size=4,gpu_memory_utilization=0.8,enforce_eager=True,max_gen_toks=256,dtype=auto), gen_kwargs: (None), limit: None, num_fewshot: None, batch_size: 4
|    Tasks     |Version|Filter|n-shot|  Metric  |   | Value  |   |Stderr |
|--------------|------:|------|-----:|----------|---|-------:|---|------:|
|hellaswag     |      1|none  |     0|acc       |↑  |  0.3862|±  | 0.0049|
|              |       |none  |     0|acc_norm  |↑  |  0.4863|±  | 0.0050|
|lambada_openai|      1|none  |     0|acc       |↑  |  0.2249|±  | 0.0058|
|              |       |none  |     0|perplexity|↓  |955.5915|±  |77.6979|

Could you share the version of your evaluation code? I updated my `lm-evaluation-harness` and added the following code to avoid additional errors.

```diff
index 76bbe81d..f100ca54 100644
--- a/lm_eval/models/vllm_causallms.py
+++ b/lm_eval/models/vllm_causallms.py
@@ -138,7 +138,14 @@ class VLLM(TemplateLM):
                 "Found 'gemma' in model name, a BOS token will be used as Gemma series models underperform without it."
             )
 
-        if parse_version(version("vllm")) >= parse_version("0.8.3"):
+        if parse_version(version("vllm")) >= parse_version("0.8.4"):
+            self.hf_chat_template = resolve_hf_chat_template(
+                model_config=self._config,
+                tokenizer=self.tokenizer,
+                chat_template=None,
+                tools=None,
+            )
+        elif parse_version(version("vllm")) >= parse_version("0.8.3"):
             self.hf_chat_template = resolve_hf_chat_template(
                 tokenizer=self.tokenizer,
                 chat_template=None,
@@ -486,7 +493,7 @@ class VLLM(TemplateLM):
             inputs = []
             ctxlens = []
             for cache_key, context_enc, continuation_enc in chunk:
-                if full_length := (context_enc + continuation_enc) >= self.max_length:
+                if full_length := len(context_enc + continuation_enc) >= self.max_length:
                     eval_logger.warning(
                         f"Context length {full_length} exceeds max length ({self.max_length}). Truncating context."
                     )
```

**Comment by llsj14** [comment]
I'm wondering why the degradation only occurred with Gemma 3. Using the same evaluation code and scripts, both Qwen2.5-72B-Instruct and Llama-4-Scout-17B-16E-Instruct produced consistent accuracy on both Hugging Face and vLLM, without any issues.

**Comment by LucasWilkinson** [comment]
I was just thinking it may have been related to: https://github.com/vllm-project/vllm/pull/17574 since Gemma uses sliding window, this commit actually made it into https://github.com/vllm-project/vllm/releases/tag/v0.8.5.post1 (which might be worth trying but if you already tested the latest commit it might be something else)

Im using:
```
lm_eval==0.4.8
```
and just the flash attention bundled with vLLM (its pinned to a commit in vLLM)

hmmm weird, I didn't have to touch the lm-eval code to get it working 🤔

**Comment by llsj14** [comment]
Ah, you're using the latest release version of lm-eval. I'll check the installation and build process again—maybe that's where the issue is.

>hmmm weird, I didn't have to touch the lm-eval code to get it working 🤔 do you have any understanding of why these changes were required?

It turns out that the `resolve_hf_chat_template` function was updated two days ago in [vLLM](https://github.com/vllm-project/vllm/commit/96722aa81da3d103675a38cfe528dd719cd9fc0a). Additionally, a recent commit introduced a bug in lm-eval just a day ago. I've submitted a [PR](https://github.com/EleutherAI/lm-evaluation-harness/pull/2972) to address this issue.

**Comment by DarkLight1337** [comment]
> It turns out that the resolve_hf_chat_template function was updated two days ago in vLLM.

That PR shouldn't affect models that already have a default chat template. Did you notice any changes in the input prompts?

**Comment by llsj14** [comment]
Yeah, it wasn't related to the accuracy results. It was just an argument issue. I just wanted to check the version of lm-eval to reproduce Lucas's results.

**Comment by linbeyoung** [comment]
Same with vllm0.8.5, I get almost zero using gemma3-12-it. The outputs are unreadable. And I try dtype='bfloat16', now it seems normal. When I use 0.8.3, I did not have to set the dtype

**Comment by DarkLight1337** [comment]
That issue was fixed recently, so you shouldn't have to set the dtype in v0.9.0

**Comment by amahfouz** [comment]
I have an issue with Gemma3 in vllm 0.8.6 as well. I ran some evals on gemma-3-12b-pt using both hugging face and VLLM and got a much lower accuracy using VLLM. Note that this only affects multimodal inputs, whereas text-only evals give the expected accuracy.

**Comment by DarkLight1337** [comment]
For gemma3, setting `VLLM_USE_V1=0` may help because the attention mechanism is simplified in V1 which doesn't fit this model well

**Comment by d-falck** [comment]
Also experiencing this issue. All Gemma 3 models are extremely inaccurate (and `VLLM_USE_V1=0` makes no difference). Would be great to see some acknowledgment/attention/a fix.

Edit: to be clear, this is on the most recent version, not just 0.8.4.

**Comment by rlrs** [comment]
Can echo this even on v0.9.1. Not sure if it's related to Blackwell or not, but Gemma models output gibberish.

**Comment by d-falck** [comment]
As per https://github.com/vllm-project/vllm/issues/20341, the issue I'm experiencing at least seems to only happen on transformers 4.53 (behaviour is fine on 4.52)
